### PR TITLE
Hts 249: schema verification

### PR DESCRIPTION
**Pull request by adwelly**

### CODE REVIEW

File: app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala [hunk 1]
```diff
@@ -84,24 +95,43 @@ class RegisterController @Inject()(val messagesApi: MessagesApi,
...
 
   def getCreateAccountHelpToSavePage: Action[AnyContent] = authorisedForHtsWithConfidence {
     implicit request ⇒
       Future.successful(Ok(views.html.register.create_account_help_to_save()))
   }
 
+  def validateOutGoingUserInfo(userInfo: Option[NSIUserInfo]): Either[String, Option[NSIUserInfo]] = {
+    val conf = app.configuration
+    userInfo match {
+      case None => Right(None)
```
Comment (MarcusQA) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: Should this be a Left?

File: app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala [hunk 2]
```diff
@@ -202,18 +232,142 @@ class RegisterController @Inject()(val messagesApi: MessagesApi,
...
+    try {
+      val report: ProcessingReport = jsonValidator.validate(schema, userInfoJson)
+      if (report.isSuccess) Right(Some(userInfo)) else Left(report.toString)
+    } catch {
+      case e: Exception => Left(e.getMessage)
+    }
+  }
+
+  private[controllers] def validateUserInfoAdditional(userInfo: NSIUserInfo): Either[String, Option[NSIUserInfo]] =
+    if (userInfo.forename.startsWith("a")) Left("FEATURE outgoing-json-validation: Forename starts with space") else Right(Some(userInfo))
```
Comment (MarcusQA) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: Should "a" be " "?

File: app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala (hunk 1 above)
Comment (adwelly) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: It turns out it does have to be a right. There is some other code that catches this situation and deals with it correctly. If I return a left here that code is never executed. Right(None) is the appropriate return.

File: app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala [hunk 3]
```diff
@@ -203,17 +230,139 @@ class RegisterController @Inject()(val messagesApi: MessagesApi,
...
+        } { nino ⇒
           Logger.info(s"Received request to get user details: redirecting to get user details using NINO $nino as authorisation code")
           Redirect(routes.RegisterController.confirmDetails(Some(nino), None, None, None).absoluteURL())
         }
       }
     }
 
+  private[controllers] def validateUserInfoAgainstSchema(userInfo: NSIUserInfo, schema: JsonNode): Either[String, Option[NSIUserInfo]] = {
+
+    val userInfoJson = JsonLoader.fromString(Json.toJson(userInfo).toString)
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: Could this throw an exception as well?

File: test/uk/gov/hmrc/helptosavefrontend/controllers/RegisterControllerSpec.scala [hunk 4]
```diff
@@ -209,6 +216,71 @@ class RegisterControllerSpec extends TestSupport {
...
+      "If the outgoing-json futureDate function detects a birth date of today it returns a right " in {
+        val today = java.time.LocalDate.now()
+        val tomorrow = today.plus(1, java.time.temporal.ChronoUnit.DAYS)
+        val futureUser = validNSIUserInfo copy (dateOfBirth = tomorrow)
+        register.validateOutGoingUserInfo(Some(futureUser)).isLeft shouldBe true
+      }
+
+      "If the outgoing-json validateOutGoingJson function detects a birth date in the future it returns a left " in {
+        val today = java.time.LocalDate.now()
+        val tomorrow = today.plus(1, java.time.temporal.ChronoUnit.DAYS)
```
Comment (AndyHWChung) [inline test/uk/gov/hmrc/helptosavefrontend/controllers/RegisterControllerSpec.scala]: Could write `today.plusDays(1L)`

**Comment by AndyHWChung** [comment]
This was actually mostly @adwelly's work - I fixed the git history and squashed everything into one commit and now it looks like my work. Sorry!

File: app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala [hunk 5]
```diff
@@ -63,7 +63,7 @@ class NSIConnectorImpl extends NSIConnector {
       .map { response ⇒
         response.status match {
           case Status.CREATED ⇒
-            Logger.info(s"Successfully created a NSI account for ${userInfo.nino}")
+            Logger.info(s"Received 201 from NSI for create account: ${userInfo.nino}")
```
Comment (tim-holbrook) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala]: Please change this to say "Received 201 from NSI, successfully created account for ${userInfo.nino}"

File: app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala [hunk 6]
```diff
@@ -102,14 +110,28 @@ class RegisterController @Inject()(val messagesApi: MessagesApi,
...
         }
-
     }
 
   def getCreateAccountHelpToSavePage: Action[AnyContent] = authorisedForHtsWithConfidence {
     implicit request ⇒
       Future.successful(Ok(views.html.register.create_account_help_to_save()))
   }
 
+  def validateOutGoingUserInfo(userInfo: Option[NSIUserInfo]): Either[String, Option[NSIUserInfo]] = {
```
Comment (tim-holbrook) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: please change this to validateCreateAccountSchema

File: app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala [hunk 7]
```diff
@@ -220,10 +242,128 @@ class RegisterController @Inject()(val messagesApi: MessagesApi,
...
+  private[controllers] def before1800(userInfo: NSIUserInfo): Either[String, Option[NSIUserInfo]] = {
+    val year = userInfo.dateOfBirth.getYear
+    if (year < 1800) Left("FEATURE outgoing-json-validation: Date of birth before 1800") else Right(Some(userInfo))
+  }
 }
 
 object RegisterController {
 
+  object JSONValidationFeature {
+    //For ICD v 1.2
```
Comment (tim-holbrook) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: ICD v1.7

File: app/uk/gov/hmrc/helptosavefrontend/util/Toggles.scala [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    def enabled(): FEATURE_THEN[A] = {
+      conf.getBoolean(s"feature-toggles.$name.enabled") match {
+        case Some(b) =>
+          logger.info("FEATURE: " + name + " enabled")
+          FEATURE_THEN(name, b, unconfiguredVal, logger)
+        case None => throw new RuntimeException(s"FEATURE($name) is not present in configuration file - misconfigured")
+      }
+    }
+
+    //TODO: def enabledWith(additional: String): FEATURE_THEN_KEY[A] = ???
```
Comment (tim-holbrook) [inline app/uk/gov/hmrc/helptosavefrontend/util/Toggles.scala]: Delete comment, thx

File: app/uk/gov/hmrc/helptosavefrontend/util/Toggles.scala [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        Logger.info("FEATURE: " + name + " executed in " + (endTime - startTime).toString + " milliseconds.")
+        Right(result)
+      } else {
+        Left(unconfiguredVal)
+      }
+    }
+  }
+  object FEATURE_THEN
+
+//  case class FEATURE_THEN_KEY[A](name: String, key: String, enabled: Boolean, hasKey: Boolean, unconfiguredVal: A)
```
Comment (tim-holbrook) [inline app/uk/gov/hmrc/helptosavefrontend/util/Toggles.scala]: Pls delete these comments

File: app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala [hunk 10]
```diff
@@ -68,6 +75,7 @@ class RegisterController @Inject()(val messagesApi: MessagesApi,
                 nino ← EitherT.fromOption[Future](maybeNino, "could not retrieve either userDetailsUrl or NINO from auth")
                 eligible ← helpToSaveService.checkEligibility(nino, authorisationCode)
                 nsiUserInfo ← toNSIUserInfo(eligible)
+                _ <- EitherT.fromEither[Future](validateCreateAccountJsonSchema(nsiUserInfo))
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: why do we have two validations in place now ,  one inside of `toNSIUserInfo(eligible)` and one in `validateCreateAccountJsonSchema` ?

**Pull request merged**